src/runtime_src/core/common/api/hw_queue.cpp

```diff
@@ -359,6 +360,14 @@ class hw_queue_impl : public command_manager::executor
   virtual std::cv_status
   wait(const xrt_core::command* cmd, size_t timeout_ms) = 0;
 
+  // Enqueue a command returning a fence that can be waited on
+  virtual fence
+  enqueue(xrt_core::command* cmd) = 0;
+
+  // Enqueue a command with dependencies
+  virtual fence
+  enqueue(xrt_core::command* cmd, const std::vector<fence>& waits) = 0;
```

Comment:
I think this is too constraining.
You could accept a range instead, to avoid systematically allocating a vector.
Of course since you are only in C++17, an object which work with std::cbegin() and `std::cend()` for example will play the concept of range.
But a generic object does not work well with dynamic polymorphism...
So at the end I do not know how to generalize as is...
Do you have a code example for the use case?

Reply:
This is purely implementation you are showing.  I do have to get from a list of run objects (application level) to internal fence_handles  I don't understand what you mean by constraining?  I will share example with you.  

There are two new APIs added at application level, both are in xrt::run.   They are marked experimental because I am not sure they are what we ultimately want.  I am not worried about the implementation you are showing, that can always change to anything more efficient you might have in mind.   Let's talk.